File: test/OAuthGitHub.Api.Unit.Test/Controllers/SignUpControllerTest.cs
using System.Threading;
using System.Threading.Tasks;
using FluentAssertions;
using MediatR;
using Microsoft.AspNetCore.Mvc;
using NUnit.Framework;
using OAuthGitHub.Api.Controllers.Shared;
using OAuthGitHub.Api.Controllers.SignUp;
using OAuthGitHub.Api.Unit.Test.Mocks;
using OAuthGitHub.Api.Unit.Test.Stubs;

namespace OAuthGitHub.Api.Unit.Test.Controllers
{
    [TestFixture]
    public class SignUpControllerTest : ControllerTest<SignUpController>
    {
        private SignUpController _controller;
        private ActionResult     _response;
        private IMediator        _mediator;

        [SetUp]
        public void SetUp()
        {
            _mediator   = SignUpMediatorMock.Mediator();
            _controller = new SignUpController(_mediator, Logger.Object);
        }

        [Test, Order(0)]
        public async Task RegisterNonExistingUser_ShouldReturnCreatedStatus()
        {
            _response =
                await _controller.SignUp(SignUpRequestStub.Request, new CancellationToken());
            _mediator.Should().HaveBeenCalledMock();

            _response.Should().BeAssignableTo(typeof(CreatedResult))
                .And.NotBeEquivalentTo(typeof(BadRequestResult));
        }

        [Test, Order(3)]
        public void RegisterNonExistingUser_ShouldReturnTheToken()
        {
            var okResponse   = _response.As<CreatedResult>();
            var authResponse = okResponse.Value.As<AuthenticationResponse>();

            authResponse.Token.Should().BeEquivalentTo(TokenStub.Token);
        }
    }
}

File: test/OAuthGitHub.Api.Unit.Test/Controllers/SignUpControllerTest.fs
namespace OAuthGitHub.Api.Unit.Test.Controllers

open System.Threading
open FluentAssertions
open FsUnit
open FSharp.Control.Tasks.V2
open Microsoft.AspNetCore.Mvc
open OAuthGitHub.Api.Controllers.Shared
open OAuthGitHub.Api.Unit.Test.Stubs.Stubs
open NUnit.Framework
open OAuthGitHub.Api.Controllers.SignUp
open OAuthGitHub.Api.Unit.Test.Mocks.SignUpMediatorMock

[<TestFixture>]
type SignUpControllerTest() =
    inherit ControllerTest<SignUpController>()
    let mutable response: ActionResult = null

    member private this.mediator = Mediator()

    member private this.controller =
        SignUpController(this.mediator, this.Logger.Object)

    [<Test; Order(0)>]
    member this.RegisterNonExistingUser_ShouldReturnCreatedStatus() =
        task {
            let! r = this.controller.SignUp(signUpRequestStub, CancellationToken.None)
            response <- r
            HaveBeenCalledMock()
            response |> should be ofExactType<CreatedResult>
        }

    [<Test; Order(3)>]
    member this.RegisterNonExistingUser_ShouldReturnTheToken() =
        response |> should not' (equal null)
        let okResponse = response.As<CreatedResult>()

        let authResponse =
            okResponse.Value.As<AuthenticationResponse>()

        authResponse.Token |> should be (sameAs tokenStub)
